src/ant/src/AntennaChecker.cc

```diff
@@ -740,6 +740,92 @@ bool AntennaChecker::checkCSR(odb::dbTechLayer* tech_layer,
   return violation;
 }
 
+bool AntennaChecker::checkRatioViolations(odb::dbTechLayer* layer,
+                                          const NodeInfo& node_info,
+                                          bool verbose,
+                                          bool report,
+                                          std::ofstream& report_file)
+{
+  bool node_has_violation = false;
+  auto par_violation = checkPAR(layer, node_info, verbose, report, report_file);
+  bool car_violation = checkCAR(layer, node_info, verbose, report, report_file);
+  node_has_violation = par_violation.first || car_violation;
+
+  if (layer->getRoutingLevel() != 0) {
+    auto psr_violation
+        = checkPSR(layer, node_info, verbose, report, report_file);
+    bool csr_violation
+        = checkCSR(layer, node_info, verbose, report, report_file);
+    node_has_violation
+        = node_has_violation || psr_violation.first || csr_violation;
+  }
```

Comment:
Does check* have a side-effect or should we skip this if node_has_violation was already set by PAR/CAR?

Reply:
We should not skip because a net can have both PAR/CAR and PSR/CSR. Skipping it will give us the wrong number of violations and make the report incorrect.